feature/experimental/p4rt/ate_tests/google_discovery_protocol_packetin_test/google_discovery_protocol_packetin_test.go

```diff
@@ -78,6 +82,175 @@ var (
+
+	ate.Traffic().Stop(t)
+}
+
+// fetchPackets reads p4rt packets sent to p4rt client.
+func fetchPackets(ctx context.Context, t *testing.T, client *p4rt_client.P4RTClient, expectNumber int) []*p4rt_client.P4RTPacketInfo {
+	t.Helper()
+	packets := []*p4rt_client.P4RTPacketInfo{}
+	for i := 0; i < expectNumber; i++ {
+		_, packet, err := client.StreamChannelGetPacket(&streamName, 0)
+		if err == io.EOF {
+			t.Logf("EOF error is seen in PacketIn.")
+			break
+		} else if err == nil {
+			if packet != nil {
+				packets = append(packets, packet)
+			}
+		} else {
+			t.Fatalf("There is error seen when receving packets. %v, %s", err, err)
+			break
+		}
+	}
+	return packets
+}
+
+// testPacketIn programs p4rt table entry and sends traffic related to GDP,
+// then validates packetin message metadata and payload.
+func testPacketIn(ctx context.Context, t *testing.T, args *testArgs) {
+	leader := args.leader
+	follower := args.follower
+
+	// Insert wbb acl entry on the DUT
+	if err := programmTableEntry(ctx, t, leader, args.packetIO, false); err != nil {
+		t.Fatalf("There is error when programming entry")
+	}
+	// Delete wbb acl entry on the device
+	defer programmTableEntry(ctx, t, leader, args.packetIO, true)
+
+	// Send GDP traffic from ATE
+	srcEndPoint := args.top.Interfaces()[atePort1.Name]
+	testTraffic(t, args.ate, args.packetIO.GetTrafficFlow(args.ate, 300, 2), srcEndPoint, 10)
+
+	packetInTests := []struct {
+		desc       string
+		client     *p4rt_client.P4RTClient
+		expectPass bool
+	}{{
+		desc:       "PacketIn to Primary Controller",
+		client:     leader,
+		expectPass: true,
+	}, {
+		desc:       "PacketIn to Secondary Controller",
+		client:     follower,
+		expectPass: false,
+	}}
+
+	for _, test := range packetInTests {
+		t.Run(test.desc, func(t *testing.T) {
+			// Extract packets from PacketIn message sent to p4rt client
+			packets := fetchPackets(ctx, t, test.client, 40)
+
+			if !test.expectPass {
+				if len(packets) > 0 {
+					t.Fatalf("Unexpected packets received.")
+				}
+			} else {
+				if len(packets) == 0 {
+					t.Fatalf("There are no packets received.")
+				}
+				t.Logf("Start to decode packet and compare with expected packets.")
+				wantPacket := args.packetIO.GetPacketTemplate()
+				for _, packet := range packets {
+					if packet != nil {
+						if wantPacket.DstMAC != nil && wantPacket.EthernetType != nil {
+							dstMac, etherType := decodePacket(t, packet.Pkt.GetPayload())
+							if dstMac != *wantPacket.DstMAC || etherType != layers.EthernetType(*wantPacket.EthernetType) {
+								t.Fatalf("Packet in PacketIn message is not matching wanted packet.")
+							}
+						}
+
+						metaData := packet.Pkt.GetMetadata()
+						for _, data := range metaData {
+							if data.GetMetadataId() == METADATA_INGRESS_PORT {
+								if string(data.GetValue()) != args.packetIO.GetIngressPort() {
+									t.Fatalf("Ingress Port Id is not matching expectation.")
+								}
+							}
```

Comment:
There are several layers of nested if and for block.

You can reduce the layers using continue for loop. For example,

Replace
			if !test.expectPass {
				if len(packets) > 0 {
					t.Fatalf("Unexpected packets received.")
				}
			} 

with

			if !test.expectPass {
				if len(packets) > 0 {
					t.Fatalf("Unexpected packets received.")
				}
                                continue
			} 

Then, else can be removed.